### PR TITLE
MODINVSTOR-576, MODINVSTOR-901: Correctly implement delete by CQL

### PR DESCRIPTION
**Pull request by julianladisch**
* Correctly implement delete by CQL for instance/holding/item (MODINVSTOR-576)
* Reject DELETE instance/holding/item without CQL query (MODINVSTOR-901)

instanceDeleteCascade.sql automatically deletes connected marc source records
to avoid additional database queries.

For the domain event (Kafka) messages the deleted record is needed.
To avoid additional queries the delete query returns the deleted record and
the instance id it belongs to.

To avoid deserializing and serializing from/to JsonObject and/or POJO the
database returns a Json String (= raw) that is directly used for the domain
event (Kafka) message. DomainEventRaw takes raw Json Strings whereas
existing DomainEvent only takes POJOs.

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json
Comment (julianladisch) [reply descriptors/ModuleDescriptor-template.json:73]: The old code allows to delete all records. It was a mistake that the old code allows to omit the query parameter. This bug causes serious issues: https://issues.folio.org/browse/MODINVSTOR-901
Therefore it is no longer allowed to omit the query parameter. This is a bug fix.

New feature: Also allow to delete only selected records using query with CQL: https://issues.folio.org/browse/MODINVSTOR-576

File: src/main/java/org/folio/rest/support/CqlUtil.java
Comment (adamdickmeiss) [inline src/main/java/org/folio/rest/support/CqlUtil.java:7]: Quite ugly to parse CQL that way. It would be better to run the parser and see if returns TRUE (the CQL parser of folio-vertx-lib simply returns null (to indicate no WHERE clause). But since the result of this is for optimization , I guess it doesn't matter.

**Review by adamdickmeiss — COMMENTED** [review]
I don't know enough about this module anymore, so I hope others will chip in.

File: src/main/java/org/folio/services/instance/InstanceService.java
Comment (kaladay) [inline src/main/java/org/folio/services/instance/InstanceService.java:130]: Should this also check that id is safe?
That is, is it possible to inject CQL here via the `id` variable?
`String id = "c1726a23-2e7c-47c6-a278-0b177d8db9b5) or (id==*)";`

File: src/test/java/org/folio/rest/support/CqlUtilTest.java
Comment (kaladay) [inline src/test/java/org/folio/rest/support/CqlUtilTest.java:12]: I'm used to seeing the Annotation next to the Class rather than next to the Import.
I recommend swapping lines 10 and 11.

File: src/main/java/org/folio/services/domainevent/ItemDomainEventPublisher.java
Comment (julianladisch) [reply src/main/java/org/folio/services/domainevent/ItemDomainEventPublisher.java:57]: If itemRaw is

    {"id": "123","barcode": "987"}

then instanceIdAndItemRaw is

    {"instanceId":"567","id": "123","barcode": "987"}

The closing brace of itemRaw is copied and becomes the closing brace of instanceidAndItemRaw.

File: src/main/java/org/folio/services/instance/InstanceService.java
Comment (julianladisch) [reply src/main/java/org/folio/services/instance/InstanceService.java]: Good catch.
CQL encoding the id string will prevent this.

File: src/test/java/org/folio/rest/support/CqlUtilTest.java
Comment (julianladisch) [reply src/test/java/org/folio/rest/support/CqlUtilTest.java]: Fixed.

File: descriptors/ModuleDescriptor-template.json
Comment (julianladisch) [reply descriptors/ModuleDescriptor-template.json]: Agreed, major version bumped.

**Review by kaladay — APPROVED** [review]

File: src/main/resources/templates/db_scripts/instanceDeleteCascade.sql
Comment (marcjohnson-kint) [inline src/main/resources/templates/db_scripts/instanceDeleteCascade.sql:9]: I think it's interesting that active changes are being made in this area.

No one should be using MARC records stored directly in inventory and this API should've been removed a long time ago.

How has it been decided to cascade the deletion rather than fail deletion when a MARC record is present?

File: src/main/java/org/folio/services/kafka/InventoryProducerRecordBuilder.java
Comment (marcjohnson-kint) [inline src/main/java/org/folio/services/kafka/InventoryProducerRecordBuilder.java:54]: Is there a better way to handle this?

I'm surprised that there isn't a mapper that writes a string as a string that can avoid this special case workaround.

**Review by marcjohnson-kint — COMMENTED** [review]
I think the timing of this is change is unfortunate, coming only 3 days before the feature freeze for 2022 R2.

I would rather significant (and especially breaking changes) weren't made this close to the release period. 

It increases the chances of delaying the module release and that issues will be discovered in the BugFest period rather than in testing activities throughout the development cycle.

If this change goes ahead please:
* make sure there are changes in all dependent modules for supporting these new interfaces before merging
* (given the changes to message publishing) contact all of the teams that maintain modules that use the Kafka messages to test that their modules still work with those messages.

File: src/test/java/org/folio/rest/api/StorageTestSuite.java
Comment (julianladisch) [reply src/test/java/org/folio/rest/api/StorageTestSuite.java]: This is to test all possibilities.

Changed to cql.allRecords=1 as requested.

File: src/test/java/org/folio/rest/api/ItemStorageTest.java
Comment (julianladisch) [reply src/test/java/org/folio/rest/api/ItemStorageTest.java]: Fixed.

File: src/test/java/org/folio/rest/api/InstanceStorageTest.java
Comment (julianladisch) [reply src/test/java/org/folio/rest/api/InstanceStorageTest.java]: Changed.

File: src/test/java/org/folio/rest/api/HoldingsStorageTest.java
Comment (julianladisch) [reply src/test/java/org/folio/rest/api/HoldingsStorageTest.java]: Replaced by SneakyThrows

File: src/test/java/org/folio/rest/api/HoldingsStorageTest.java
Comment (julianladisch) [reply src/test/java/org/folio/rest/api/HoldingsStorageTest.java]: I disagree, but changed it to get this done.

File: src/main/java/org/folio/services/kafka/InventoryProducerRecordBuilder.java
Comment (julianladisch) [reply src/main/java/org/folio/services/kafka/InventoryProducerRecordBuilder.java:54]: I couldn't find such a mapper.

File: src/main/java/org/folio/services/item/ItemService.java
Comment (julianladisch) [reply src/main/java/org/folio/services/item/ItemService.java]: Because multiple nested lambdas and expressions are more readable when curly braces show code stand on its own.

Changed to get this done.

File: src/main/java/org/folio/services/instance/InstanceService.java
Comment (julianladisch) [reply src/main/java/org/folio/services/instance/InstanceService.java]: The code also deletes the MARC record itself. This is the information it is intended to provide.

File: NEWS.md
Comment (julianladisch) [reply NEWS.md:8]: Done.

File: src/main/resources/templates/db_scripts/instanceDeleteCascade.sql
Comment (julianladisch) [reply src/main/resources/templates/db_scripts/instanceDeleteCascade.sql:9]: You should have asked this question before you merged the pull request four years ago:
https://github.com/folio-org/mod-inventory-storage/pull/126/files#diff-dbf6259d56a78ce52a3e91870b73432e7ac83f8b06f58a22135feb02fb17a9bdR349

File: src/test/java/org/folio/rest/api/InstanceDomainEventTest.java
Comment (julianladisch) [reply src/test/java/org/folio/rest/api/InstanceDomainEventTest.java:87]: All four test methods in this class check a failure caused by a foreign key violation:
eventIsNotSentWhenUpdateFailed()
eventIsNotSentWhenCreateFailed()
eventIsNotSentWhenRemoveFailed()
eventIsNotSentWhenRemoveAllFailed()
They all expect a 400 HTTP status code.
Can you give an example how "Failed" could be reworded to better reflect the expected failure?

File: src/main/java/org/folio/rest/support/CqlQuery.java
Comment (adamdickmeiss) [inline src/main/java/org/folio/rest/support/CqlQuery.java:11]: Any value for cql.allRecords should match.

File: src/main/java/org/folio/rest/support/CqlQuery.java
Comment (adamdickmeiss) [inline src/main/java/org/folio/rest/support/CqlQuery.java:28]: I don't like this. Parsing CQL (and SQL) by regular expressions is just a bad idea. I would have preferring letting the CQL2PgJSON parser do the work and if it evaluates to `true` or `TRUE`, it would match. Admitted, that would not work with id= searches.. But that's a very lame way of getting all records anyway.

For example dealing with `(cql.allRecords = ok)` as well.

**Review by adamdickmeiss — COMMENTED** [review]
I don't have enough knowledge to judge either way about this.

File: src/main/java/org/folio/rest/support/CqlQuery.java
Comment (marcjohnson-kint) [reply src/main/java/org/folio/rest/support/CqlQuery.java:28]: > I don't like this. Parsing CQL (and SQL) by regular expressions is just a bad idea. I would have preferring letting the CQL2PgJSON parser do the work and if it evaluates to `true` or `TRUE`, it would match. 

I agree that would be better.

> Admitted, that would not work with id= searches.. But that's a very lame way of getting all records anyway.

I completely agree with this. I don't think we should try to be clever about figuring out if a query matches all records.

Comment (julianladisch) [reply src/main/java/org/folio/rest/support/CqlQuery.java:44]: The regex pattern matching has been replaced with CQLParser.

If you want the single DELETE_ALL Kafka event been replaced by potentially thousands of single record DELETE Kafka events please create a Jira and I will be handle it in a separate pull request. Replacing DELETE_ALL by multiple DELETEs will be a performance regression that requires approval by the sysops.

File: src/main/resources/templates/db_scripts/instanceDeleteCascade.sql
Comment (julianladisch) [reply src/main/resources/templates/db_scripts/instanceDeleteCascade.sql:9]: I couldn't find a Jira that approves the removal of the MARC records APIs.

File: src/test/java/org/folio/rest/api/InstanceDomainEventTest.java
Comment (julianladisch) [reply src/test/java/org/folio/rest/api/InstanceDomainEventTest.java:87]: Resolving because the intention of this test is to demonstrate a foreign key violation, and not an internal failure.

File: src/main/java/org/folio/rest/support/CqlQuery.java
Comment (marcjohnson-kint) [reply src/main/java/org/folio/rest/support/CqlQuery.java:44]: > Replacing DELETE_ALL by multiple DELETEs will be a performance regression that requires approval by the sysops.

How would this be a regression? As I understand it, the semantically explicit `cql.allRecords` publishes the single message.

Have I misunderstood?

File: src/main/resources/templates/db_scripts/instanceDeleteCascade.sql
Comment (marcjohnson-kint) [reply src/main/resources/templates/db_scripts/instanceDeleteCascade.sql:9]: There may well not be one. 

I would hope it was obvious that it has been superseded by mod-source-record-storage.

I thought those APIs were informally deprecated a long time ago.

File: src/main/java/org/folio/rest/support/CqlQuery.java
Comment (julianladisch) [reply src/main/java/org/folio/rest/support/CqlQuery.java:44]: You wrote:

> I don't think we should try to be clever about figuring out if a query matches all records.

If we don't figure out whether the query matches all records (for example `cql.allRecords`) we cannot call the fast deleteAllInstances() and always need to call the slow delete(String cql) method. This is a performance regression.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/rest/support/CqlQuery.java:44]: > You wrote:
> 
> > I don't think we should try to be clever about figuring out if a query matches all records.
> 
> If we don't figure out whether the query matches all records (for example `cql.allRecords`) we cannot call the fast deleteAllInstances() and always need to call the slow delete(String cql) method. This is a performance regression.

I believe in my other [comment|https://github.com/folio-org/mod-inventory-storage/pull/775#discussion_r886650100] I suggested that I thought that a dedicated message was acceptable for the case where the intent was obvious and explicit (cql.allRecords).

My objection was not to that optimisation when the intention was obvious, it was with the inference of intent with other queries that happened to delete all records.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/rest/support/CqlQuery.java:44]: To be very explicit, I think treating cql.allRecords as an explicit way to deliberately delete all records is a good idea and in that case only a specific message could be published.

I don't think that any special provision should be taken for other queries that only happen to delete all records. 

It seems that code is still attempting to do that.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/rest/support/CqlQuery.java:44]: I wouldn't consider accommodating only the explicit intent of `cql.allRecords` a regression. This is the replacement for the current API.

I do think the code attempting to infer intent (listed below) about other queries should be removed. I would not consider that a regression, because that functionality did not exist previously.

```
    return "id".equalsIgnoreCase(node.getIndex())
        && StringUtils.equalsAny(base, "=", "==")
        && "*".equals(node.getTerm());
```

Comment (julianladisch) [reply src/main/java/org/folio/rest/support/CqlQuery.java:44]: Since RMB v30.2.3 the section ["CQL: Matching all records" of the RMB documentation](https://github.com/folio-org/raml-module-builder#cql-matching-all-records) declares `id=*` to be officially supported:

> A search matching all records in the target index can be executed with a cql.allRecords=1 (CQL standard) or a id=* (RMB specific) query.

Should we add a restriction to the RMB documentation and the mod-inventory-storage documentation and the Nolana release notes saying that mod-inventory-storage no longer has full support for `id=*` in all cases?

Comment (marcjohnson-kint) [reply src/main/java/org/folio/rest/support/CqlQuery.java:44]: > Should we add a restriction to the RMB documentation and the mod-inventory-storage documentation and the Nolana release notes saying that mod-inventory-storage no longer has full support for id=* in all cases?

I think @adamdickmeiss summed that up pretty well:

> Admitted, that would not work with id= searches.. But that's a very lame way of getting all records anyway.

I think it's poor guidance to recommend that folks use a query like `id=*` for matching all records, especially when there is a specific provision for an intent revealing clause in the standard.

I ambivalent about whether mod-inventory-storage allows that CQL or not. My concern here is with implementing specific additional behaviour based upon it. I already think that the special message is a compromise.

I don't think this decision needs to be made for all cases, rather the case that we are exploring here, which is a specific behaviour for deletion only.

There are lots of CQL queries that folks can write that don't make much sense. That is one of the major trade offs with using an open query language rather than defined criteria.

If you want to include documentation to describe how folks should avoid using `id=*` etc for matching all records and for deletion specifically, I would support that endeavour.

**Review by steveellis — APPROVED** [review]

**Pull request merged**
